### PR TITLE
WELD-2427 WeldInstance#select and WeldEvent#select with Type.

### PR DESCRIPTION
**Pull request by manovotn**

### CODE REVIEW

File: impl/src/main/java/org/jboss/weld/resolution/ResolvableBuilder.java [hunk 1]
```diff
@@ -50,7 +51,8 @@
 
 public class ResolvableBuilder {
 
-    private static final Class<?>[] FACADE_TYPES = new Class<?>[] { Event.class, Instance.class, WeldInstance.class, Provider.class, InterceptionFactory.class };
+    private static final Class<?>[] FACADE_TYPES = new Class<?>[] { Event.class, Instance.class, WeldEvent.class,
```
Comment (manovotn) [inline impl/src/main/java/org/jboss/weld/resolution/ResolvableBuilder.java]: Not really sure we need to change this?

File: environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ *
+ * @author <a href="mailto:manovotn@redhat.com">Matej Novotny</a>
+ */
+@RunWith(Arquillian.class)
+public class WeldEventTest {
```
Comment (manovotn) [inline environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java]: The tests are verifying:
* ability to `@Inject WeldEvent<X>`
* known functionality when replacing `Event` with `WeldEvent`
* new selections based on `Type` and expected exceptions

I am accepting additional ideas for corner cases and/or possible problematic areas.

File: impl/src/main/java/org/jboss/weld/resolution/ResolvableBuilder.java (hunk 1 above)
Comment (mkouba) [reply impl/src/main/java/org/jboss/weld/resolution/ResolvableBuilder.java]: I believe we need this.

File: environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            try {
+                // using Type from List<Object> -> List<?>
+                bean.getEventListObject().select(listWildcardType);
+                Assert.fail();
+            } catch (IllegalStateException e) {
+                // expected
+            }
+
+            // verify valid behaviour, following selections should work
+            bean.getEventObject().select(firstType);
```
Comment (mkouba) [inline environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java]: We should probably also test observability here... i.e. that a child `Event` works as expected?

File: impl/src/main/java/org/jboss/weld/bean/builtin/InstanceImpl.java [hunk 4]
```diff
@@ -151,9 +151,15 @@ public boolean isUnsatisfied() {
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public <X> WeldInstance<X> select(Type subtype, Annotation... qualifiers) {
-        // TODO Auto-generated method stub
-        return null;
+        // verify if this was invoked on WeldInstance<Object>
+        if (!this.getType().getTypeName().equals(Object.class.getName())) {
```
Comment (mkouba) [inline impl/src/main/java/org/jboss/weld/bean/builtin/InstanceImpl.java]: `!this.getType().equals(Object.class)` should be fine...

File: environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/instance/enhanced/selectType/WeldInstanceSelectTypeTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    public static Archive<?> createTestArchive() {
+        return ClassPath.builder().add(ShrinkWrap.create(BeanArchive.class, Utils.getDeploymentNameAsHash(WeldInstanceSelectTypeTest.class))
+            .addPackage(WeldInstanceSelectTypeTest.class.getPackage())).build();
+    }
+
+    @Test
+    public void testValidSelect() {
+        try (WeldContainer container = new Weld().initialize()) {
+            // Class<T> implements Type, so let's make use of that
+            Type firstType = SomeInterface.class;
```
Comment (mkouba) [inline environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/instance/enhanced/selectType/WeldInstanceSelectTypeTest.java]: Maybe we should add some tests for parameterized types...

File: environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/instance/enhanced/selectType/WeldInstanceSelectTypeTest.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            try {
+                someInterfaceInstance.select(secondType);
+                Assert.fail("Selecting by type from any other type than Object should fail with IllegalStateException!");
+            } catch (IllegalStateException e) {
+                // expected
+            }
+        }
+    }
+    
+    // TODO add tests which will use a @Typed bean eliminating Object as bean type
```
Comment (mkouba) [inline environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/instance/enhanced/selectType/WeldInstanceSelectTypeTest.java]: I would also add a test that `WeldInstance.destroy()` works if you create a new `@Dependent` bean instance with the new `select()` method.

File: environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/instance/enhanced/selectType/WeldInstanceSelectTypeTest.java (hunk 5 above)
Comment (manovotn) [reply environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/instance/enhanced/selectType/WeldInstanceSelectTypeTest.java]: Hmm, can add that too. Though there isn't much of a change here (as opposed to WeldEvent where I have such test). You could have done the same before with WeldInstance.

File: impl/src/main/java/org/jboss/weld/bean/builtin/InstanceImpl.java (hunk 4 above)
Comment (manovotn) [reply impl/src/main/java/org/jboss/weld/bean/builtin/InstanceImpl.java]: Hmm yep, it eluded me yet again that Class is in fact Type :)

File: impl/src/main/java/org/jboss/weld/resolution/ResolvableBuilder.java (hunk 1 above)
Comment (manovotn) [reply impl/src/main/java/org/jboss/weld/resolution/ResolvableBuilder.java]: I _think_ it's for when you resolve bean via BM (why would you do that with Event though?), but I am not really sure. If I remove it `@Inject` will keep working.

File: environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java (hunk 3 above)
Comment (manovotn) [reply environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java]: See `testEventObservability` method

Comment (mkouba) [reply environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java]: Ah, you're right ;-)

**Comment by manovotn** [comment]
Added tests for `@PreDestroy` and `@Typed` and one simple valid select for `ParameterizedType`

**Pull request merged**